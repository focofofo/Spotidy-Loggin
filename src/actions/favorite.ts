import { FavoritePayload } from '../models'

export const SET_FAVORITES = 'SET_FAVORITES'
export const REMOVE_FAVORITE = 'REMOVE_FAVORITE'
export const RESET_FAVORITES = 'RESET_FAVORITES'

/**
 * Save the favorite list into the storage
 * @param payload favorite list result of the API
 */
export function doSetFavorites(payload: FavoritePayload) {
  return {
    type: SET_FAVORITES,
    payload,
  }
}

/**
 * Removes the favorite track from the storage given the track index
 * @param payload index of the track to be removed
 */
export function doRemoveFavorite(payload: number) {
  return {
    type: REMOVE_FAVORITE,
    payload,
  }
}

/**
 * Clear the favorite storage
 */
export function doResetFavorites() {
  return {
    type: RESET_FAVORITES,
  }
}
