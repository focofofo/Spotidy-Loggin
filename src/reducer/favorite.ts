import { AnyAction } from 'redux'
import {
  REMOVE_FAVORITE,
  RESET_FAVORITES,
  SET_FAVORITES,
} from '../actions/favorite'
import { FavoritePayload } from '../models'

const initialState: FavoritePayload = {
  href: '',
  items: [],
  limit: 0,
  next: '',
  offset: 0,
  previous: '',
  total: 0,
}

export default function favorite(
  state: FavoritePayload = initialState,
  action: AnyAction
) {
  switch (action.type) {
    case SET_FAVORITES:
      return {
        ...state,
        ...action.payload,
      }
    case REMOVE_FAVORITE:
      state.items.splice(action.payload, 1)
      return {
        ...state,
        total: state.total - 1,
      }
    case RESET_FAVORITES:
      return {
        ...initialState,
      }
    default:
      return state
  }
}
