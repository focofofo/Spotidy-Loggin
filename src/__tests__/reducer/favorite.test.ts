import { AnyAction } from 'redux'
import favorite from '../../reducer/favorite'
import {
  SET_FAVORITES,
  REMOVE_FAVORITE,
  RESET_FAVORITES,
} from '../../actions/favorite'
import MockFavorite from '../mock/Favorite.mock.json'

describe('favorite reducer', () => {
  it('should handle initial state', () => {
    expect(favorite(undefined, {} as AnyAction)).toMatchSnapshot()
  })

  it('should handle SET_FAVORITES', () => {
    expect(
      favorite(undefined, {
        type: SET_FAVORITES,
        payload: {
          href: 'test',
          items: [],
          limit: 10,
          next: 'test',
          offset: 10,
          previous: 'test',
          total: 10,
        },
      })
    ).toMatchSnapshot()
  })

  it('should handle REMOVE_FAVORITE', () => {
    expect(
      favorite(
        {
          href: 'test',
          items: [MockFavorite],
          limit: 10,
          next: 'test',
          offset: 10,
          previous: 'test',
          total: 10,
        },
        {
          type: REMOVE_FAVORITE,
          payload: 0,
        }
      )
    ).toMatchSnapshot()
  })

  it('should handle RESET_FAVORITES', () => {
    expect(favorite(undefined, { type: RESET_FAVORITES })).toMatchSnapshot()
  })
})
