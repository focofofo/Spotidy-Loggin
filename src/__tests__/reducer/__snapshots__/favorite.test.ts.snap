// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`favorite reducer should handle REMOVE_FAVORITE 1`] = `
Object {
  "href": "test",
  "items": Array [],
  "limit": 10,
  "next": "test",
  "offset": 10,
  "previous": "test",
  "total": 9,
}
`;

exports[`favorite reducer should handle RESET_FAVORITES 1`] = `
Object {
  "href": "",
  "items": Array [],
  "limit": 0,
  "next": "",
  "offset": 0,
  "previous": "",
  "total": 0,
}
`;

exports[`favorite reducer should handle SET_FAVORITES 1`] = `
Object {
  "href": "test",
  "items": Array [],
  "limit": 10,
  "next": "test",
  "offset": 10,
  "previous": "test",
  "total": 10,
}
`;

exports[`favorite reducer should handle initial state 1`] = `
Object {
  "href": "",
  "items": Array [],
  "limit": 0,
  "next": "",
  "offset": 0,
  "previous": "",
  "total": 0,
}
`;
